### PR TITLE
askeladd: round11 — lion_beta2=0.999 (optimizer sweep, single delta)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Lion's `β2` parameter controls the exponential moving average of the **squared-gradient surrogate** (the sign term's denominator is not explicitly squared in Lion, but β2 still controls the momentum buffer for the momentum correction). SOTA uses `lion_beta2=0.99` (default). Raising to `β2=0.999` makes the optimizer "remember" further back — slower adaptation to gradient direction changes but potentially smoother convergence.

This is a **single-delta** test: one hyperparameter changed from `0.99` → `0.999`, all else SOTA-identical.

**Why β2=0.999 might help:** With the cosine LR schedule absent (SOTA uses `lr_cosine_t_max=0`), the LR is constant at 1e-4. A longer-memory β2 could smooth the gradient landscape in Lion's sign-based updates, reducing per-step noise while still converging — analogous to AdamW's β2 sensitivity, where higher β2 works better with flat LR schedules.

**Risk:** Slower adaptation could hurt early convergence (ep0–2), with potential recovery by ep9. Lion is known to be sensitive to β1; β2 has not been tested on tay.

## Instructions

Single change from SOTA: add `--lion-beta2 0.999` to the SOTA command.

**Exact command:**
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.999 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --use-ema --ema-decay 0.999 \
  --wandb-group tay-round11-lion-beta2-0p999 \
  --wandb-name askeladd/round11-lion-beta2-0p999
```

This is the **only** change vs SOTA `d03oghpp`: `--lion-beta2 0.999` (SOTA default is 0.99).

Note: Do **not** add `--lr-cosine-t-max` — SOTA `d03oghpp` used `lr_cosine_t_max=0` (flat LR).

## Baseline

Current SOTA — PR #115 (thorfinn, lr=1e-4 + EMA=0.999), test_abupt **10.580**, val 9.484 (ep9), W&B run `d03oghpp`.

| Metric | PR #115 SOTA | AB-UPT ref |
|---|---:|---:|
| `abupt` | **10.580** | — |
| `surface_pressure` | 5.690 | 3.82 |
| `wall_shear` | 10.419 | 7.29 |
| `volume_pressure` | 12.740 | 6.08 |
| `tau_x` | 8.908 | 5.35 |
| `tau_y` | 12.491 | 3.65 |
| `tau_z` | 13.071 | 3.63 |

**Reference val trajectory (SOTA):** 53.75 / 24.15 / 16.51 / 13.47 / 11.83 / 10.88 / 10.16 / 9.73 / 9.48

## Results

_(fill in after training completes)_

Post: best `val_primary/abupt_axis_mean_rel_l2_pct`, final `test_primary/abupt_axis_mean_rel_l2_pct`, per-component test breakdown, and W&B run URL.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Results

**Verdict: hypothesis falsified.** Lion `β2=0.999` monotonically slowed convergence vs SOTA `β2=0.99`. The "slower-adaptation-but-smoother-convergence" trade did not pay off within the 9-epoch budget — it was strictly worse at every epoch from ep1 onward, with no recovery.

### Headline metrics

| Metric | This run (β2=0.999) | SOTA β2=0.99 (PR #115) | Δ vs SOTA |
|---|---:|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` (best, ep9) | **11.493** | 9.484 | **+21.2%** (worse) |
| `test_primary/abupt_axis_mean_rel_l2_pct` | **12.564** | 10.580 | **+18.8%** (worse) |

### Test per-component breakdown (ep9 EMA checkpoint, 50 cases)

| Component | This run | SOTA #115 | AB-UPT ref | Δ vs SOTA |
|---|---:|---:|---:|---:|
| `surface_pressure_rel_l2_pct` | 7.173 | 5.690 | 3.82 | +26.1% |
| `wall_shear_rel_l2_pct` | 12.574 | 10.419 | 7.29 | +20.7% |
| `volume_pressure_rel_l2_pct` | 14.159 | 12.740 | 6.08 | +11.1% |
| `wall_shear_x_rel_l2_pct` (tau_x) | 10.832 | 8.908 | 5.35 | +21.6% |
| `wall_shear_y_rel_l2_pct` (tau_y) | 14.945 | 12.491 | 3.65 | +19.6% |
| `wall_shear_z_rel_l2_pct` (tau_z) | 15.713 | 13.071 | 3.63 | +20.2% |

MAEs: surface_p_mae=0.0207, wall_shear_mae=0.1185, volume_p_mae=31.36.

### Val trajectory (per-epoch `val_primary/abupt_axis_mean_rel_l2_pct`)

| Epoch | β2=0.999 (this run) | β2=0.99 (SOTA `d03oghpp`) |
|---:|---:|---:|
| 0 | 55.42 | 53.75 |
| 1 | 31.77 | 24.15 |
| 2 | 22.44 | 16.51 |
| 3 | 18.44 | 13.47 |
| 4 | 15.63 | 11.83 |
| 5 | 13.89 | 10.88 |
| 6 | 12.78 | 10.16 |
| 7 | 11.92 | 9.73 |
| 8 | 11.49 | 9.48 |

ep0 is comparable, then β2=0.999 falls behind from ep1 and the gap (~2 pct points) persists through ep8 with no sign of catching up.

### Run metadata

- W&B run (rank0): [`tfumujfi`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/tfumujfi) (group `tay-round11-lion-beta2-0p999`)
- Peak GPU memory (rank0): ~74.2 GB / 80 GB (HBM3, ~72.3% util)
- Total wall time: 270.8 min (8×H100, DDP)
- Best epoch: 9 (selected by EMA val abupt)
- Train loss converged smoothly; gradients stable, zero clipped after warmup, no nonfinite grads.

### Exact command

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.999 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --use-ema --ema-decay 0.999 \
  --wandb-group tay-round11-lion-beta2-0p999 \
  --wandb-name askeladd/round11-lion-beta2-0p999
```

(No code changes — single CLI delta vs SOTA `d03oghpp`.)

### What happened

The hypothesis posited that a longer-memory β2 might smooth Lion's sign-based updates with the constant-LR schedule (`lr_cosine_t_max=0`). Instead, β2=0.999 strictly degraded convergence:

- **Mechanistic read**: In Lion, β2 controls the momentum buffer that is mixed with the current gradient before sign extraction (the update direction is `sign(β1·m + (1-β1)·g)` and the buffer is updated with `m ← β2·m + (1-β2)·g`). Higher β2 means the buffer carries stale gradients longer. With a constant LR and only 9 effective epochs, that stale momentum dominates the early-training direction and the optimizer takes longer to "turn" toward the true loss minimum. Sign-based updates amplify this effect because direction matters more than magnitude.
- **Behavioral read**: ep0 is virtually identical (init noise dominates), but every subsequent epoch underperforms by a roughly constant margin — the trajectory is shape-preserving, not just shifted. No phase change, no late recovery. This is consistent with "slow exploration, no benefit from smoothing in this regime" rather than "trade short-term cost for long-term gain."
- **No instability indicators**: gradient norms, clipping, and nonfinite counts all behave like SOTA. So the slowdown is a learning-rate-effective issue, not stability/numerics.

The risk flagged in the PR ("Slower adaptation could hurt early convergence (ep0–2), with potential recovery by ep9") materialized in the first half but the second half never recovered.

### Suggested follow-ups

1. **β2 sweep at low end**: try `β2 ∈ {0.9, 0.95, 0.98}` to see whether *shorter* memory helps with constant LR — the symmetric prediction of this result.
2. **Re-test β2=0.999 with cosine schedule** (`--lr-cosine-t-max 9`): the failure mode may be regime-specific to flat LR; warm-down could let stale momentum turn into late-stage noise reduction. Lower priority than (1).
3. **Test β1 sensitivity** (e.g., `β1 ∈ {0.85, 0.95}`) — more impactful on Lion per the original Lion paper, and untested on tay.
4. **Longer run with β2=0.999**: 9 epochs may be too short for the smoothing benefit to manifest; if ep budget grows (e.g., 20+ ep), revisit. Probably not worth it given the consistent ~21% gap.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #161 because ADVISOR: Experiment complete — lion_beta2=0.999 is a clear dead end. W&B run tfumujfi shows best_val=11.493 (vs SOTA 9.484, +21.2% regression) and test_abupt=12.564 (vs SOTA 10.580, +18.7% regression). Reducing beta2 from 0.99→0.999 makes Lion behave more like Adam (more momentum smoothing), which kills the fast-convergence advantage Lion has on this task. The full val trajectory confirms it never converges competitively. Closing. Askeladd, no results comment needed — the W&B data tells the full story..
